### PR TITLE
Address build delays

### LINKED ISSUE #916: Default branch build gaps after throttling change
Analysis deletes versions on changes and we lose the associated builds. Since we now delay default branch builds until the next day, we've created a gap in the build matrix and show "?" until the builds kick in.

I believe we've got the following options to address this

1. Roll back change to build trigger
2. Introduce additional tracking fields to support latest version display on package page during cool-down period and keep version with builds
3. Introduce a new “intraday latest branch” version for the purpose of the package page display that is without builds but won’t get built until the cool-down period has expired. Then roll over once that build has happened.
4. Roll back build throttling and introduce a longer version update window during analysis for branch changes

I think there are only two non-hacky ways to do this:

- Option 1 - roll back and pay the build price
- Option 4 - roll back build throttle + longer version update window

I’m leaning towards 4. The way this would work is that during analysis we ignore changes to default branch versions within X hours, say 8 or 12. That way, any changes that happen within that window get collated into a single refresh, such that we’d build at most 2 or 3 times per day while also not falling too far behind with the “last updated” display on the website. It shouldn’t really matter if the latest change (for just the default branch) is 8h or 12h behind.

We'd pick up a few more builds than we do with current throttling (unless we decide to make the window 24h and accept the update metadata on the package page to be more outdated).

The upside is that the throttling query becomes simpler again, albeit at the expense of more complex analysis version handling. OTOH, that's outside of an SQL query and probably easier to test and understand.

### PR DESCRIPTION
**Pull request by finestructure**
Fixes #916

### CODE REVIEW

File: Sources/App/Commands/Analyze.swift [hunk 1]
```diff
@@ -143,6 +143,9 @@ func analyze(client: Client,
                          threadPool: threadPool,
                          transaction: tx,
                          packages: packages)
+                .map { throttleBranchVersions(
+                    packageDeltas: $0,
+                    delay: Constants.branchVersionRefreshDelay) }
```
Comment (finestructure) [inline Sources/App/Commands/Analyze.swift:148]: This is the core of it: after determining the `VersionDelta` (versions to be added, deleted, kept), we go through the three arrays and remove any recently changes `.branch` versions from `toAdd` and move the existing ones from `toDelete` to `toKeep`, essentially undoing any branch changes detected if they happened within the configured time window.

File: Sources/App/Commands/Analyze.swift [hunk 2]
```diff
@@ -354,21 +364,17 @@ func diffVersions(client: Client,
...
         return transaction.eventLoop.future(error: AppError.invalidPackageCachePath(package.id, package.url))
     }
     guard let pkgId = package.id else {
         return transaction.eventLoop.future(error: AppError.genericError(nil, "PANIC: package id nil for package \(package.url)"))
     }
-    
-    let defaultBranch = Repository.defaultBranch(on: transaction, for: package)
-        .map { b -> [Reference] in
-            if let b = b { return [.branch(b)] } else { return [] }  // drop nil default branch
-        }
```
Comment (finestructure) [inline Sources/App/Commands/Analyze.swift:370]: This was an unnecessary query we ran against the `repositories` table. We were already joining against it. I dropped the whole `Repository.defaultBranch` function.

File: Sources/App/Commands/Analyze.swift [hunk 3]
```diff
@@ -399,16 +406,51 @@ func diffVersions(client: Client,
...
+    let rejectRecentBranchVersions: (Version) -> Bool = {
+        !selectRecentBranchVersions($0)
+    }
+
+    return .init(
+        toAdd: deltas.toAdd.filter(rejectRecentBranchVersions),
+        toDelete: deltas.toDelete.filter(rejectRecentBranchVersions),
+        toKeep: deltas.toKeep + deltas.toDelete.filter(selectRecentBranchVersions)
+    )
+}
```
Comment (finestructure) [inline Sources/App/Commands/Analyze.swift:439]: This implements the "throttling"

File: Sources/App/Models/Version.swift [hunk 4]
```diff
@@ -113,7 +113,13 @@ final class Version: Model, Content {
 
 extension Version: Equatable {
     static func == (lhs: Version, rhs: Version) -> Bool {
-        lhs.id == rhs.id
+        if let id1 = lhs.id, let id2 = rhs.id {
+            return id1 == id2
+        } else {
+            return lhs.commit == rhs.commit
+                && lhs.commitDate == rhs.commitDate
+                && lhs.reference == rhs.reference
```
Comment (finestructure) [inline Sources/App/Models/Version.swift:121]: This was a nasty issue for our test assertions: there were several cases where we simply created versions like so without saving:

```
let v = Version(package: p)
```

i.e. without `save()` and without assigning `id:` which meant we were comparing different versions with `nil` ids...

Comment (finestructure) [reply Sources/App/Models/Version.swift:121]: (I don't think it was any existing cases but new ones I just wrote for this PR)

File: Sources/App/Commands/Analyze.swift [hunk 5]
```diff
@@ -371,23 +406,44 @@ func diffVersions(client: Client,
...
+    }
+
+    guard let incomingVersion = incoming.latestBranchVersion,
+          let latestIncoming = incomingVersion.commitDate else {
+        // there's no incoming branch version -> leave incoming alone (which will lead to removal)
+        return incoming
+    }
+
+    #warning("distance to latestIncoming to to Current.date() ?")
+    let delta = latestIncoming.timeIntervalSinceReferenceDate - latestExisting.timeIntervalSinceReferenceDate
```
Comment (finestructure) [inline Sources/App/Commands/Analyze.swift:438]: In practise this should make much of a difference as we're discovering new versions every couple of hours at most. I.e. there won't really be a meaningful difference between when a new version appears and `Current.date()`

I think we should go with `Current.date()`.

Comment (finestructure) [reply Sources/App/Commands/Analyze.swift:438]: As discussed, change to `Current.date()`

**Review by daveverwer — APPROVED** [review]

File: Sources/App/Commands/Analyze.swift (hunk 5 above)
Comment (finestructure) [reply Sources/App/Commands/Analyze.swift]: Fixed

**Pull request merged**